astroid/interpreter/_import/util.py

```diff
@@ -18,18 +22,44 @@ def is_namespace(modname: str) -> bool:
     # That's unacceptable here, so we fallback to _find_spec_from_path(), which does
     # not, but requires instead that each single parent ('astroid', 'nodes', etc.)
     # be specced from left to right.
-    components = modname.split(".")
-    for i in range(1, len(components) + 1):
-        working_modname = ".".join(components[:i])
+    processed_components = []
+    last_submodule_search_locations: _NamespacePath | None = None
+    for component in modname.split("."):
+        processed_components.append(component)
+        working_modname = ".".join(processed_components)
         try:
-            # Search under the highest package name
-            # Only relevant if package not already on sys.path
-            # See https://github.com/python/cpython/issues/89754 for reasoning
-            # Otherwise can raise bare KeyError: https://github.com/python/cpython/issues/93334
-            found_spec = _find_spec_from_path(working_modname, components[0])
+            # the path search is not recursive, so that's why we are iterating
+            # and using the last parent path
+            found_spec = _find_spec_from_path(
+                working_modname, last_submodule_search_locations
+            )
         except ValueError:
             # executed .pth files may not have __spec__
             return True
+        except KeyError:
+            # Intermediate steps might raise KeyErrors
+            # https://github.com/python/cpython/issues/93334
+            # TODO: update if fixed in importlib
+            # For tree a > b > c.py
+            # >>> from importlib.machinery import PathFinder
+            # >>> PathFinder.find_spec('a.b', ['a'])
+            # KeyError: 'a'
+
+            # Repair last_submodule_search_locations
+            if last_submodule_search_locations:
+                # TODO: py38: remove except
+                try:
+                    # pylint: disable=unsubscriptable-object
+                    last_item = last_submodule_search_locations[-1]
+                except TypeError:
+                    last_item = last_submodule_search_locations._recalculate()[-1]
+                assumed_location = pathlib.Path(last_item) / component
```

Comment:
```suggestion
                # e.g. for failure example above, add 'a/b' and keep going
                # so that find_spec('a.b.c', path=['a', 'a/b']) succeeds
                assumed_location = pathlib.Path(last_item) / component
```